File: app/src/main/java/re/snapwi/orientation/MainActivity.java
package re.snapwi.orientation;

import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.text.Html;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.ViewFlipper;
import javax.inject.Inject;
import re.snapwi.orientation.io.Joke;

import static android.view.View.OnClickListener;
import static re.snapwi.orientation.JokePresenter.JokeListener;
import static re.snapwi.orientation.JokePresenter.JokeModule;

public class MainActivity extends ActionBarActivity implements OnClickListener, JokeListener {
  private static final String FLIPPER = "FLipperStateKey";

  private ViewFlipper flipper;
  private TextView jokeTextView;

  @Inject JokePresenter presenter;

  @Override protected void onCreate(Bundle savedState) {
    super.onCreate(savedState);
    setContentView(R.layout.activity_main);

    ((App) getApplication()).appComponent().plus(new JokeModule(this)).inject(this);

    jokeTextView = (TextView) findViewById(R.id.jokeTextView);
    flipper = (ViewFlipper) findViewById(R.id.mainFlipper);
    findViewById(R.id.tryAgainButton).setOnClickListener(this);

    if (savedState != null) {
      flipper.setDisplayedChild(savedState.getInt(FLIPPER, 0));
    }

    presenter.restoreState(savedState);
  }

  @Override protected void onResume() {
    super.onResume();
    presenter.onResume();
  }

  @Override protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putInt(FLIPPER, flipper.getDisplayedChild());
    presenter.saveState(outState);
  }

  @Override protected void onPause() {
    presenter.onPause();
    super.onPause();
  }

  @Override public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.menu_main, menu);
    return true;
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    if (item.getItemId() == R.id.action_refresh) {
      presenter.getRandomJoke();
      return true;
    }

    return super.onOptionsItemSelected(item);
  }

  @Override public void onClick(View v) {
    presenter.getRandomJoke();
  }

  @Override public void onJokeLoaded(Joke joke) {
    jokeTextView.setText(Html.fromHtml(joke.getJoke()));
    flipper.setDisplayedChild(2);
  }

  @Override public void onFailed() {
    flipper.setDisplayedChild(1);
  }

  @Override public void onStartedLoading() {
    flipper.setDisplayedChild(0);
  }
}

File: app/src/main/kotlin/re/snapwi/orientation/ui/MainActivity.kt
package re.snapwi.orientation.ui

import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.text.Html
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import butterknife.bindView
import re.snapwi.orientation.R
import re.snapwi.orientation.di.JokeModule
import re.snapwi.orientation.io.Joke
import re.snapwi.orientation.util.getAppComponent
import javax.inject.Inject
import kotlin.properties.Delegates

public class MainActivity : AppCompatActivity(), JokePresenter.JokeListener, SwipeRefreshLayout.OnRefreshListener {
  val refreshLayout: SwipeRefreshLayout by bindView(R.id.swipeRefreshLayout)
  val retryLayout: View by bindView(R.id.retryLayout)
  val jokeTextView: TextView by bindView(R.id.jokeTextView)

  var presenter: JokePresenter by Delegates.notNull()
    @Inject set

  override fun onCreate(state: Bundle?) {
    super<AppCompatActivity>.onCreate(state)
    setContentView(R.layout.activity_main)
    getAppComponent().plus(JokeModule(this)).inject(this)

    findViewById(R.id.tryAgainButton).setOnClickListener({ presenter.getRandomJoke() })

    refreshLayout.setOnRefreshListener(this)
    refreshLayout.setRefreshing(if (state == null) true else state.getBoolean(LOADING_CONTENT))
    presenter.restoreState(state ?: Bundle())
  }

  override fun onRefresh() {
    presenter.getRandomJoke()
  }

  override fun onResume() {
    super<AppCompatActivity>.onResume()
    presenter.onResume()
  }

  override fun onSaveInstanceState(outState: Bundle) {
    super<AppCompatActivity>.onSaveInstanceState(outState)
    outState.putBoolean(LOADING_CONTENT, refreshLayout.isRefreshing())
    presenter.saveState(outState)
  }

  override fun onPause() {
    presenter.onPause()
    super<AppCompatActivity>.onPause()
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    getMenuInflater().inflate(R.menu.menu_main, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    if (item.getItemId() == R.id.action_refresh) {
      presenter.getRandomJoke()
      return true
    }

    return super<AppCompatActivity>.onOptionsItemSelected(item)
  }

  override fun onJokeLoaded(joke: Joke) {
    jokeTextView.setText(Html.fromHtml(joke.getJoke()))
    refreshLayout.setRefreshing(false)
    retryLayout.setVisibility(View.INVISIBLE)
  }

  override fun onFailed() {
    refreshLayout.setRefreshing(false)
    retryLayout.setVisibility(View.VISIBLE)
  }

  override fun onStartedLoading() {
    refreshLayout.setRefreshing(true)
    retryLayout.setVisibility(View.INVISIBLE)
  }

  companion object {
    private val LOADING_CONTENT = "FlipperStateKey"
  }
}